feat: new auth adapter

Moumouls opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
Allow user to set up webauthn (biometric authentication on mobile device) once he is logged, and then i can log in with webauthn (fingerprint scan on mobile device). Rework auth adapter too support MFA systems, auth challenges systems and better interfaces when writing an auth adapter.

Related issue: #7042 

### Approach
Implement new WebAuthn Adapter with a challenge JWT strategy to avoid to store challenge into the db.
Add new auth hook methods tiggered in dedicated use cases.

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete suggested TODOs that do not apply to this PR.
-->

- [x] Add tests
- [x] Add changes to documentation (guides, repository pages, in-code descriptions)
- [x] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [x] Add new Parse Error codes to Parse JS SDK <!-- no hard-coded error codes in Parse Server -->
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)


Futur work: Need to implement client side webauthn system in JS SDK for an easy usage.

Moumouls commented:
WebAuthn adapter finished, Parse Server is ready to handle FIDO and biometric authentication (mbp, mobile devices) 🚀

davimacedo commented:
Great job @Moumouls ! @dplewis @mtrezza @dblythy any additional thoughts here?

Moumouls commented:
Okay in real world implementation of an OTP SMS system i detected wrong implementation on challenge handling, i will push the fix then we will be okay !

Moumouls commented:
Okay i pushed some fixes after some beta testing on my company project that use extensively the Auth System.

i implemented an SMS OTP adapter, everything works perfectly.

OTP Adapter pseudo code

```ts
export const OTPAuth = {
	policy: 'additional',

	async challenge(
		challengeData: boolean,
		authData: OtpAuthDataInput,
		options: any,
		req: any,
		user?: Parse.User,
	) {
		if (!user || !user.get('phone')) throw new Error('User not found')
		const otp = new OTP(user.id, user.get('phone'))
		await SMSAdapter.send(
			user.get('phone'),
			`OTP code : ${await otp.generate()}`,
		)
	},

	async validateSetUp(authData: CreateUpdateOtpAuthDataInput, options: any, req: AuthReq) {
		return checkAuthorization(req)
	},

	async validateUpdate(authData: CreateUpdateOtpAuthDataInput, options: any, req: AuthReq) {
		return checkAuthorization(req)
	},

	async validateLogin(authData: OtpAuthDataInput, options: any, req: AuthReq, user?: Parse.User) {
		try {
			// As an additional auth system user should be already identified
			// with another default auth system
			if (!authData.code || !user?.id) return throwError()
			if (!user.get('phone')) return throwError()
			const otp = new OTP(user.id, user.get('phone'))
			await otp.check(authData.code)
			return { doNotSave: true }
		} catch (e) {
			// eslint-disable-next-line @typescript-eslint/no-throw-literal
			throw new Parse.Error(403, e.message)
		}
	},
}
```

Moumouls commented:
@davimacedo @mtrezza @dplewis , so this one is now fully finished and also tested on a critical security project !

Moumouls commented:
closed https://github.com/parse-community/parse-server/pull/7052 in favor of this one

davimacedo commented:
@Moumouls it looks the tests are failing after the last commit

Moumouls commented:
Thanks @davimacedo , I'll take look this morning. Do you think we can merge this one asap ?

davimacedo commented on spec/ParseUser.spec.js:
```diff
@@ -1789,7 +1789,7 @@ describe('Parse.User testing', () => {
   });
 
-  it('should allow login with old authData token', done => {
+  it('should not allow login with old authData token', async () => {
```
@mtrezza @dplewis should we change this? Any other comment for this PR?

mtrezza replied:
Maybe we can pick up the discussion in the advisory, I think there were some open questions what it means to change this?

davimacedo replied:
Should we add that as a breaking change?

dblythy commented:
Looks good so far @Moumouls!

Would you see any value in adding a helper method to `Parse.Cloud` that can register auth, rather then it all being required in `index.js`?

For example, I have all my login code in `/functions/login.js`.

It would be nice to be able to write:

```javascript
Parse.Cloud.registerAuth('authName', {...});
```

Rather than having to export and import back to index.js.

Moumouls commented:
So here after a master update, how can help to make this one merged ? :)

mtrezza commented:
Resolve the open comments. There is an open discussion about stale auth data in the advisory.

Moumouls commented:
@mtrezza @davimacedo  this pr is now synced with master, but it seems that CI have issue with mongodb runner, any idea on this ?

mtrezza commented on CHANGELOG.md:
```diff
@@ -91,7 +91,9 @@ ___
 ### Breaking Changes
 - Improved schema caching through database real-time hooks. Reduces DB queries, decreases Parse Query execution time and fixes a potential schema memory leak. If multiple Parse Server instances connect to the same DB (for example behind a load balancer), set the [Parse Server Option](https://parseplatform.org/parse-server/api/master/ParseServerOptions.html) `databaseOptions.enableSchemaHooks: tru ...
 - Added file upload restriction. File upload is now only allowed for authenticated users by default for improved security. To allow file upload also for Anonymous Users or Public, set the `fileUpload` parameter in the [Parse Server Options](https://parseplatform.org/parse-server/api/master/ParseServerOptions.html) (dblythy, Manuel Trezza) [#7071](https://github.com/parse-community/parse-server/pu ...
+- Auth Adapter now always validate authData at login tim (Moumouls) [#7079](https://github.com/parse-community/parse-server/pull/7079)
```
Could you clarify what this means? Not sure developers would know whether/how this change is affecting their deployment and how to address this change?

mtrezza commented:
Not sure what the issue here is, but looking at other PRs like https://github.com/parse-community/parse-server/pull/7312 the CI works (expect for some nasty flaky tests)

mtrezza commented on package.json:
```diff
@@ -23,10 +23,12 @@
     "@parse/push-adapter": "3.4.0",
     "@parse/s3-files-adapter": "1.6.0",
     "@parse/simple-mailgun-adapter": "1.2.0",
+    "@simplewebauthn/server": "^2.2.0",
```
Remove range operator

Moumouls replied:
After a long search on the web, implementing a webauthn server (and also a front) handler on our side... will be a nightmare. This lib is well coded by Masterkale (i contributed to this pkg also for some minor fixes). Webauthn follow a fixed W3C spec; currently Masterkale and some other contributors keep it synced correctly.

In node/npm eco system from my point of view this the best lib to implement this feature correctly with a maximal device compatibility.

mtrezza commented:
@Moumouls Could you try to delete package-lock and regenerate the file? The error looks like git can't install from a remote repo, likely because the repo URI points is outdated.

Moumouls commented:
Okay we should be on green here now 🚀  @mtrezza

Moumouls commented:
Okay @mtrezza  i found that since npm 7.0  (node 15) `npm i ` rewrite the lock to v2 (i've this versions on my setup). I think we could add a new CI check with a simple regex to ensure that Parse Server lock is not changed/rewrited since it seems that it leads to weird issues 🙂  What do you think about this ? @dplewis may be you have a suggestion for this one ?

Moumouls commented:
SDK (optional PR) synced: https://github.com/parse-community/Parse-SDK-JS/pull/1276

mtrezza commented:
@Moumouls Good idea, we currently have no CI check for package-lock. I will look into adding one.

mtrezza commented:
I think we should pick this up and merge in the next days. This is open for quite some time already.

@Moumouls From the changelog I understand that there are no breaking changes from this PR, correct?

Moumouls commented:
Yes @mtrezza no breaking changes, just some new interfaces and the security patch on authData validation.

Moumouls commented:
i'll try to find some time next/end of the week to resolve conflicts and adjust last details

Moumouls commented:
We should definitely merge this one asap

mtrezza commented:
Great, thanks, I'll be glad to see this merged.

Moumouls commented:
Hi @mtrezza, i've some hard time to work on this I will try to do my best to re-sync with master

mtrezza commented:
Changing to draft; PR needs rebase and resolving open questions in previous review feedback

mtrezza commented:
> ## ⚠️ Important change for merging PRs from Parse Server 5.0 onwards!
  >
  > We are planning to release the first beta version of Parse Server 5.0 in October 2021.
  >
  > If a PR contains a breaking change and is not merged before the _beta_ release of Parse Server 5.0, it cannot be merged until the end of 2022. Instead it has to follow the [Deprecation Policy](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#deprecation-policy) and phase-in breaking changes to be merged during the course of 2022.
  >
  > One of the most voiced community feedbacks was the demand for predictability in breaking changes to make it easy to upgrade Parse Server. We have made a first step towards this by introducing the Deprecation Policy in February 2021 that assists to phase-in breaking changes, giving developers time to adapt. We will follow-up with the introduction of [Release Automation](https://github.com/parse-community/parse-server/issues/7271) and a branch model that will allow breaking changes only with a new major release, scheduled for the beginning of each calendar year.
  >
  > We understand that some PRs are a long time in the making and we very much appreciate your contribution. We want to make it easy for PRs that contain a breaking change and were created _before_ the introduction of the Deprecation Policy. These PRs can be merged with a breaking change without being phased-in before the beta release of Parse Server 5.0. We are making this exception because we appreciate that this is a time of transition that requires additional effort from contributors to adapt. We encourage everyone to prepare their PRs until the end of September and account for review time and possible adaptions.
  > 
  > If a PR contains a breaking change and should be merged before the beta release, please mention @parse-community/server-maintenance and we will coordinate with you to merge the PR.
  >
  > *Thanks for your contribution and support during this transition to Parse Server release automation!*

mtrezza commented:
@Moumouls Do you think you can schedule some time anytime soon to go over this PR and get it ready for merge? I understand from the previous comments that it doesn't include a breaking change, is that still correct?

I'm afraid, the longer this PR is stale, the more difficult it will become to merge it.

Moumouls commented:
@mtrezza I'll take a look ☺️

Moumouls commented:
@mtrezza  branch synced, squashed, prettier up to date,  Webauthn and Auth spec split done. Ready to review ? 🥳 

Some flacky tests again. I'll take a look tomorrow to code cov issue.

mtrezza commented:
Wow, super quick! Let's review this!

Moumouls commented:
okay @mtrezza after a morning of hunting uncovered lines and adding some news test, minor refactor we are now fully in green here

There is some uncovered lines on webauthn verified error, it's hard to have a `verified === false`, for safety the uncovered line just throw. 


EDIT:
Oh i missed some lines on src/RestWrite.js

EDIT2:
Okay now my PR is fully covered

davimacedo commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2140,3 +2152,990 @@ describe('facebook limited auth adapter', () => {
+      { sessionToken: user.getSessionToken() }
+    );
+
+    expect(user2.get('authDataResponse')).toEqual({ baseAdapter2: { someData2: true } });
```
Can you explain this one?

Moumouls replied:
Here we check that authDataResponse works also when the user is logged in and send some authData (here baseAdapter2)
` user2.id = user.id;` and `{ sessionToken: user.getSessionToken() }`

davimacedo reviewed (commented):
@Moumouls sorry for the delay. That's a great job and I believe that's ready to be merged. I left two comments. Please take a look.

Moumouls commented:
@davimacedo  
> Should we add that as a breaking change?

Since this feature will land on Parse 5 i think we can flag it as a breaking change. @mtrezza  what is your suggestion ?

mtrezza commented:
Could you elaborate what exactly this will break and what a developer has to do to adapt their server and/or client apps?

davimacedo commented:
The breaking change would be potentially for an app that for any reason is caching an authentication provider token and using this same token to login to Parse Server multiple times, even when the token is already expired. Probably there isn't many apps relying on that but it is a breaking change. @Moumouls do you think that would be possible to have an option to switch between the old and new behaviors, according to our [deprecation policy](https://github.com/parse-community/parse-server/blob/alpha/CONTRIBUTING.md#deprecation-policy)?

Moumouls commented:
Hi @davimacedo 

Your explanation is right, but here I'm not sure if a deprecation policy should be used in this case since we have here a user account bad practice which introduce then a user account security problem.

Developers will be notified in the change log or blog of parse V5 , if this PR is okay for Parse V5

davimacedo commented:
I think that the PR is ready to be merged and go with Parse 5 once we've sorted this breaking change out. I agree with you that it will only break apps which are badly using the generated token, but it is still a breaking change that can make some apps stop working. That's why I believe that the optional behavior (if not too much effort to add that) would be good. From what I understood on the new deprecation policy, I think we must add that even for this case, right @mtrezza ?

Moumouls commented:
@davimacedo  are you interested about pushing the deprecation switch ? 🙂 

I don't think I'll be able to free up the time to work on the patch, December is going to be pretty busy for my company. So I would need some help to get through the line before the end of December.

mtrezza commented:
> From what I understood on the new deprecation policy, I think we must add that even for this case, right

That's correct. We have released the 5-beta for developers to be able to adapt their deployments for Parse Server 5. No further breaking changes should go into Parse Server 5 anymore at this point, according to our policy.

davimacedo commented:
@Moumouls I can help on that. Could you please give me access to the branch?

Moumouls commented:
Not sure how to give you access to the branch ? @davimacedo 
Do you have a github doc on this ?

You can also open a PR to my branch on my fork if you want 🙂  i'll merge it

davimacedo commented:
You'd actually need to give me access to the whole repository. But no problem, I will open a PR to your branch and let you lnow.

Moumouls commented:
@mtrezza  do you think we have a chance to get this one into the v5 release ?

davimacedo commented:
@Moumouls I've just sent you a PR with the switch: https://github.com/Moumouls/parse-server/pull/3

Let's publish it to 5.0

Moumouls commented:
@mtrezza @davimacedo we are ready here 🚀

davimacedo reviewed (dismissed):
It looks good to me. LGTM!

Moumouls commented:
@mtrezza  if you have an idea for the failing node 12/14 CI 🙂

Moumouls commented:
Hi @dblythy, we should have this one merged may be this month (if you have an ETA in mind @mtrezza ) then i can help you to implement an MFA adapter :) MFA could be a great concrete example for the documentation.

dblythy commented:
I would be happy to have a look @Moumouls

mtrezza commented on DEPRECATIONS.md:
```diff
@@ -8,6 +8,7 @@ The following is a list of deprecations, according to the [Deprecation Policy](h
 | DEPPS2 | Config option `directAccess` defaults to `true` | [#6636](https://github.com/parse-community/parse-server/pull/6636)   | 5.0.0 (2022)                    | 6.0.0 (2023)                    | deprecated            | -     |
 | DEPPS3 | Config option `enforcePrivateUsers` defaults to `true` | [#7319](https://github.com/parse-community/parse-server/pull/7319)   | 5.0.0 (2022)                    | 6.0.0 (2023)                    | deprecated            | -     |
 | DEPPS4 | Remove convenience method for http request `Parse.Cloud.httpRequest` | [#7589](https://github.com/parse-community/parse-server/pull/7589)   | 5.0.0 (2022)                    | 6.0.0 (2023)                    | deprecated            | -     |
+| DEPPS5 | Allow login with old authData token | [#7079](https://github.com/parse-community/parse-server/pull/7079)   | 5.0.0 (2022)                    | 6.0.0 (2023)                    | deprecated            | -     |
```
Can we rename this more specifically as `expired authData token`?

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2158,3 +2172,991 @@ describe('facebook limited auth adapter', () => {
+      // more than one result should never occur in Parse
+      // but a developer with a direct access to the database
+      // could break something, login/register will be canceled
+      // to avoid non determinist login between users found by the authData query
```
can you use equal line length in this comment block to make it easier to read, also rephrase, it's hard to understand without punctuation.

Here is my try, but I still don't fully understand the comment:


> Second call is triggered after `beforeSave` if `authData` was touched by the developer [what does "touched by dev" mean?] in Cloud Code.
> More than one result should never occur in Parse [what does that mean?], but a developer with direct access to the database could break something [break what and how?].
> The login or sign-up request will fail to avoid non-deterministic login between users [what is a "non-deterministic login between users"?] found by the authData query.
> true, true, just simulate a query return with length === 2 [what does that mean?]

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2158,3 +2172,991 @@ describe('facebook limited auth adapter', () => {
+    expect(firstCall[2].auth).toBeDefined();
+    expect(firstCall[3]).toBeUndefined();
+
+    // Use request sine JS SDK not ready
```
Remove comment, unless it's necessary?

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2158,3 +2172,991 @@ describe('facebook limited auth adapter', () => {
+  it('should pass authData, options, req, user to validateAuthData', async () => {
+    spyOn(baseAdapter, 'validateAuthData').and.resolveTo({});
+    await reconfigureServer({ auth: { baseAdapter } });
+
```
remove empty lines

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2158,3 +2172,991 @@ describe('facebook limited auth adapter', () => {
+      modernAdapter2: { id: 'modernAdapter2' },
+    });
+  });
+  it('should trigger correctly validateLogin', async () => {
```
add empty line before each test `it(...)`

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2158,3 +2172,991 @@ describe('facebook limited auth adapter', () => {
+    await reconfigureServer({
+      auth: { baseAdapter, baseAdapter2 },
+    });
+    await reconfigureServer({
```
is this 2x on purpose?

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2158,3 +2172,991 @@ describe('facebook limited auth adapter', () => {
+    expect(user.getSessionToken()).toBeDefined();
+    // Should not re validate the baseAdapter
+    // when user is already logged in and authData not changed
+    await user.save(
```
Can you move these into it's on sub-`describe` and instead of adding a comment to describe the test, add it as a test `it('should not re-validate...')`. You can add a `describe` as a child of another `describe` if necessary.

Fix description typos:
`Should not re-validate the baseAdapter when master key is used and authData has not changed`

Always add a empty line between tests (`it(...)`)

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -13,6 +13,63 @@ const passwordCrypto = require('../lib/password');
 const Config = require('../lib/Config');
 const cryptoUtils = require('../lib/cryptoUtils');
 
+describe('allowOldAuthDataToken option', () => {
```
I suggest to rename this to `allowExpiredAuthDataToken`, because "old" can be ambiguous

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -2260,37 +2342,14 @@ describe('Parse.User testing', () => {
     });
 
     describe('anonymous users', () => {
-      beforeEach(() => {
```
why was this removed?

Moumouls replied:
there is only 1 test under this describe. And i also improved test to easily show the "should not fail on case insensitive matches"

We only call the logIn 2 times, this big array is not necessary.

mtrezza commented on src/Adapters/Auth/index.js:
```diff
@@ -59,21 +61,63 @@ const providers = {
+      typeof adapter.validateLogin === 'function' &&
+      typeof adapter.validateUpdate === 'function'
+    ) {
+      // We can consider for DX purpose when masterKey is detected, we should
```
What does this comment mean? Maybe remove "We can consider for DX purpose"?

mtrezza commented on src/Adapters/Auth/index.js:
```diff
@@ -59,21 +61,63 @@ const providers = {
+      }
+
+      if (isLoggedIn) {
+        // User is currently updating his authData
```
// User is updating their authData

mtrezza commented on src/Adapters/Auth/index.js:
```diff
@@ -59,21 +61,63 @@ const providers = {
+        return adapter.validateSetUp(authData, options, req, user);
+      }
+
+      // Not logged in and authData configured into the DB
```
"configured into the DB" -> what does that mean? Maybe "authData is stored in DB"?

mtrezza commented on src/Adapters/Auth/index.js:
```diff
@@ -59,21 +61,63 @@ const providers = {
-      return Promise.resolve();
-    });
+
+      // Finally, user not logged in and the provider is not setup
```
```
// User not logged in and the provider is not set up, for example when a new user
// signs up or an existing user uses a new auth provider

```

mtrezza commented on src/Adapters/Auth/index.js:
```diff
@@ -59,21 +61,63 @@ const providers = {
 }
 
 function loadAuthAdapter(provider, authOptions) {
+  // `providers` are default providers already implemented
```
// `providers` are auth providers implemented by default

mtrezza commented on src/Adapters/Auth/index.js:
```diff
@@ -59,21 +61,63 @@ const providers = {
+  // `providers` are default providers already implemented
+  // into parse-server
   let defaultAdapter = providers[provider];
+  // authOptions could contain a complete custom adapter
```
// `authOptions` can contain complete custom auth adapters or
// a default auth adapter like Facebook

mtrezza commented on src/Adapters/Storage/Postgres/PostgresStorageAdapter.js:
```diff
@@ -1308,15 +1307,20 @@ export class PostgresStorageAdapter implements StorageAdapter {
         object['authData'][provider] = object[fieldName];
         delete object[fieldName];
         fieldName = 'authData';
+        // Avoid pushing authData multiple times to the
```
// Avoid adding `authData` multiple times to the query

mtrezza commented on src/Auth.js:
```diff
@@ -83,7 +91,8 @@ const getAuthForSessionToken = async function ({
       include: 'user',
     };
-
+    // For cyclic dep
```
What does that mean? If this is to avoid dependency cycle, you can remove the comment, we have a CI check for that; same with the comments below

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +313,159 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+      provider && provider.adapter && provider.adapter.policy === 'solo' && authData[provider.name]
+  );
+
+  // Solo providers can be considered as safe
```
What is a "solo provider"? Maybe add an explanation?
Condense comment block by making comment lines longer

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +313,159 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+      if (authData[provider.name]) {
+        return true;
+      } else {
+        // Push missing provider for plausible error return
```
What does that mean, "plausible error return"? Maybe rephrase comment

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +313,159 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+  let user;
+  if (foundUser) {
+    user = Parse.User.fromJSON({ className: '_User', ...foundUser });
+    // Find the user by session and current object id
```
// Find user by session and current objectId; only pass user if it's the current user or master key is provided

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +313,159 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+    await user.fetch({ useMasterKey: true });
+  }
+
+  // Perform validation as step by step pipeline
```
I think this comment can be removed, it doesn't seem to provide any substantial info

Moumouls replied:
i prefer to keep this comment it explain why this part should not be a parallel code

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +313,159 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+  // for better error consistency and also to avoid to trigger a provider (like OTP SMS)
+  // if another one fail
+  return reducePromise(
+    // apply sort to run the pipeline each time in the same order
```
remove empty line

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +313,159 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+        if (!Object.keys(validationResult).length) acc.authData[provider] = authData[provider];
+
+        if (validationResult.response) acc.authDataResponse[provider] = validationResult.response;
+        // Some auth providers after initialization will avoid
```
condense comment into 1 line

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +313,159 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+          acc.authData[provider] = validationResult.save || authData[provider];
+        }
+      } else {
+        // Support current authData behavior
```
condense comment into 1 line

mtrezza commented on src/GraphQL/loaders/usersMutations.js:
```diff
@@ -355,6 +379,57 @@ const load = parseGraphQLSchema => {
+  const challengeMutation = mutationWithClientMutationId({
+    name: 'Challenge',
+    description:
+      'The challenge mutation can be used to initiate an authentication challenge when an auth adapter need it.',
```
"needs it"

mtrezza commented on src/GraphQL/loaders/usersMutations.js:
```diff
@@ -355,6 +379,57 @@ const load = parseGraphQLSchema => {
+      },
+      authData: {
+        description:
+          'Auth data allow to pre identify the user if the auth adapter need pre identification.',
```
"allows to preidentify ... needs preidentification"

mtrezza commented on src/GraphQL/loaders/usersMutations.js:
```diff
@@ -355,6 +379,57 @@ const load = parseGraphQLSchema => {
+      },
+      challengeData: {
+        description:
+          'Challenge data payload, could be used to post some data to auth providers if they need data for the response.',
```
"can be used to post data to auth providers"

mtrezza commented on src/Options/docs.js:
```diff
@@ -67,6 +68,7 @@
  * @property {String} restAPIKey Key for REST calls
  * @property {Boolean} revokeSessionOnPasswordReset When a user changes their password, either through the reset password email or while logged in, all sessions are revoked if this is true. Set to false if you don't want to revoke sessions.
  * @property {Boolean} scheduledPush Configuration for push scheduling, defaults to false.
+ * @property {SchemaOptions} schema Rest representation on Parse.Schema https://docs.parseplatform.org/rest/guide/#adding-a-schema
```
Why does this PR add SchemaOptions?

Moumouls replied:
no idea aahaha

mtrezza replied:
Please remove if this is not auto-generated by `npm run definitions`. There is an issue with SchemaOptions https://github.com/parse-community/parse-server/issues/7805 which is a separate issue from this PR.

mtrezza commented on src/Routers/UsersRouter.js:
```diff
@@ -174,14 +175,24 @@ export class UsersRouter extends ClassesRouter {
   async handleLogIn(req) {
     const user = await this._authenticateUserFromRequest(req);
+    const authData = req.body && req.body.authData;
+    // Check if user has provided his required auth providers
```
"provided their required"

mtrezza commented on src/Routers/UsersRouter.js:
```diff
@@ -228,6 +239,17 @@ export class UsersRouter extends ClassesRouter {
       req.config
     );
 
+    // If we have some new validated authData
```
Condense comment into 1 line

mtrezza commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +479,97 @@ export class UsersRouter extends ClassesRouter {
+  async handleChallenge(req) {
+    const { username, email, password, authData, challengeData } = req.body;
+
+    // if username or email provided with password try to find the user with default
```
Condense comment into 1 line; what does "default system" mean?

mtrezza commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +479,97 @@ export class UsersRouter extends ClassesRouter {
+    if (authData) {
+      if (typeof authData !== 'object')
+        throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'authData should be an object.');
+      // To avoid security issue we should only support one identifying method
```
Is this comment necessary?; What does it mean "we should"? Do we already do it, or should we do it in a future PR, or should the developer consider this?

Moumouls replied:
yes we can remove it, the error is sufficient.

mtrezza commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +479,97 @@ export class UsersRouter extends ClassesRouter {
+        // Find the provider used to find the user
+        const provider = Object.keys(authData).find(key => authData[key].id);
+
+        // Validate authData used to identify the user
```
condense comment into 1 line; what is a "guess ID attack"?

Moumouls replied:
a guess attack (could be on email, open id, ...) allow an attacker to know easily if a targeted user exists into the DB.
This kind of attack cloud be limited by validating the authData (id + access token). So the challenge endpoint will not be subject to guess attacks.

mtrezza commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +479,97 @@ export class UsersRouter extends ClassesRouter {
+      }
+    }
+
+    // Execute challenge step by step
```
// Execute challenge with consistent order

Can you briefly explain in the comment why this is relevant?

mtrezza reviewed (changes requested):
Looks good, mostly style and typos feedback, then we should be good to merge

Moumouls commented:
thanks for the review @mtrezza , i'll check this asap. Do you think that this PR could be released under a v5.1 ?

mtrezza commented:
The release version is not determined by me but by the order of PR merging.

Moumouls commented:
@mtrezza normally, i resolved all your comment, i also squashed the branch.
i noticed also 2 things:

- Prettier is not added to lint and lint fix so we have some inconsistency on formatting. I can suggest to now use lint staged.
- i encountered a weird error with eslint cache option, i needed to remove the option and re add the option to be able to commit (since lint-fix failed)

I hope we can merge this one now :) 

EDIT: i removed my changes related to prettier since it's out of the scope of this pr

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2160,3 +2172,1033 @@ describe('facebook limited auth adapter', () => {
+        i++;
+        return originalFn(...params);
+      }
+      // Second call is triggered after beforeSave.
```
can you use equal line length in this comment block to make it easier to read, also rephrase, it's hard to understand without punctuation.

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +309,154 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+      provider && provider.adapter && provider.adapter.policy === 'solo' && authData[provider.name]
+  );
+
+  // Solo providers can be considered as safe, so we do not have to check if the user need
```
"if the user needs"

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +309,154 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+
+  // Solo providers can be considered as safe, so we do not have to check if the user need
+  // to provide an additional provider to login. An auth adapter with "solo" (like webauthn) means
+  // that no "additional" auth need be provided to login (like OTP, MFA)
```
"no additional authentication needs to be provided"

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +309,154 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+    await user.fetch({ useMasterKey: true });
+  }
+
+  // Perform validation as step by step pipeline for better error consistency
```
"step-by-step"

mtrezza commented on src/Auth.js:
```diff
@@ -298,11 +309,154 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+  }
+
+  // Perform validation as step by step pipeline for better error consistency
+  // and also to avoid to trigger a provider (like OTP SMS) if another one fail
```
"step-by-step" ... "if another one fails"

mtrezza commented on src/Options/Definitions.js:
```diff
@@ -28,6 +28,12 @@ module.exports.ParseServerOptions = {
     action: parsers.arrayParser,
   },
+  allowExpiredAuthDataToken: {
+    env: 'PARSE_SERVER_ALLOW_OLD_AUTH_DATA_TOKEN',
```
This still says "OLD" token in `env`. Please run `npm run definitions` to make sure this is corrected automatically.

mtrezza commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +478,94 @@ export class UsersRouter extends ClassesRouter {
+        // Find the provider used to find the user
+        const provider = Object.keys(authData).find(key => authData[key].id);
+
+        // Validate authData used to identify the user to avoid guess id attack
```
"to avoid brute-force attack on \`id\`"

mtrezza commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +478,94 @@ export class UsersRouter extends ClassesRouter {
+    }
+
+    // Execute challenge step by step with consistent order for better error feedback
+    // and to avoid to trigger others challenges if one of them fail
```
"step-by-step" ... "one of them fails"

mtrezza reviewed (changes requested):
Last typos, then we should be good

dblythy commented on spec/ParseUser.spec.js:
```diff
@@ -13,6 +13,63 @@ const passwordCrypto = require('../lib/password');
+        .filter(
+          log =>
+            log.args[0] ===
+            `DeprecationWarning: The Parse Server option 'allowExpiredAuthDataToken' default will change to 'false' in a future version.`
```
Can we prefer `expect(logSpy.warn).toHaveBeenCalledWith(`?

Moumouls replied:
we should not, this system will not fail if we add new deprecation policies in the future. If we use haveBeenCalledWith, future contributors to deprecation warnings could start to make this test fail.

dblythy replied:
feel free to resolve 👍

dblythy commented on spec/ParseUser.spec.js:
```diff
@@ -13,6 +13,63 @@ const passwordCrypto = require('../lib/password');
+  it('should default true', async () => {
+    const logger = require('../lib/logger').logger;
+    const logSpy = spyOn(logger, 'warn').and.callFake(() => { });
+    await reconfigureServer({});
```
Is this line necessary?

Moumouls replied:
yes it's required to reset the server config, the test above modify the config: `await reconfigureServer({ allowExpiredAuthDataToken: false });`

dblythy commented on spec/ParseUser.spec.js:
```diff
@@ -13,6 +13,63 @@ const passwordCrypto = require('../lib/password');
+  it('should enforce boolean values', async () => {
+    const options = [[], 'a', '', 0, 1, {}, 'true', 'false'];
+    for (const option of options) {
+      await expectAsync(reconfigureServer({ allowExpiredAuthDataToken: option })).toBeRejected();
```
can we use `Promise.all` here for faster test speeds

Moumouls replied:
you can't make this in parallel since here we reconfigure the server in each for loop

dblythy replied:
Ahh yep

dblythy commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,47 @@ describe('Parse.User testing', () => {
+    defaultConfiguration.auth.shortLivedAuth.setValidAccessToken('otherToken');
+    try {
+      await Parse.User._logInWith('shortLivedAuth', {});
+      fail('should not authenticate');
```
`expectAsync().toBeRejected()`

dblythy commented on spec/WebauthnAuthenticationAdapter.spec.js:
```diff
@@ -0,0 +1,407 @@
+    expect(options.user).toEqual({ id: user.id, name: 'username', displayName: 'test@test.test' });
+
+    await reconfigureServer({
+      publicServerURL: 'https://subdomain.a-super-app.com/parse',
```
Should we enforce `example.parseplatform.org/parse` or something for placeholder urls? cc @mtrezza

mtrezza replied:
Good catch, we should be using `example.com` according to [RFC 6761](https://datatracker.ietf.org/doc/html/rfc6761#section-6.5):

> The domains "example.", "example.com.", "example.net.",
   "example.org.", and any names falling within those domains, are
   special in the following ways:
>
>   1.  Users SHOULD understand that example names are reserved for use
       in documentation.

dblythy commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
    */
-  validateAppId(appIds, authData, options) {
+  validateAppId(appIds, authData, options, req) {
+    return Promise.resolve({});
```
Since this is an upgrade of the Auth Adapters, can we remove the requirement to `return Promise.resolve({})`?

Moumouls replied:
@dblythy this now supported. Do you have a suggestion about this specific line ?

I think it still relevant to keep the Prose return to show that validateAppId support promise.

mtrezza replied:
@dblythy can this be resolved?

mtrezza replied:
@dblythy can this be resolved?

dblythy commented on src/Adapters/Storage/Postgres/PostgresStorageAdapter.js:
```diff
@@ -1308,15 +1307,19 @@ export class PostgresStorageAdapter implements StorageAdapter {
         delete object[fieldName];
         fieldName = 'authData';
+        // Avoid adding authData multiple times to the query
+        if (authDataAlreadyExists) return;
```
style, prefer:

```
if (authDataAlreadyExists) {
  return;
}
```

dblythy commented on src/Auth.js:
```diff
@@ -1,5 +1,13 @@
-const RestQuery = require('./RestQuery');
```
What's the reason behind removing this and then `require`ing it in line 3 other times?

Moumouls replied:
Circular dependency. Auth.js is used into RestQuery. So the RestQuery module should be loaded dynamically when needed.

dblythy commented on src/Auth.js:
```diff
@@ -298,11 +309,154 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+    findPromise = config.database.find('_User', { $or: query }, { limit: 2 });
+  }
+
+  return findPromise;
```
Perhaps `return query.length === 0 ? Promise.resolve([]) : config.database.find('_User', { $or: query }, { limit: 2 })`

dblythy commented on src/Auth.js:
```diff
@@ -298,11 +309,154 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+  // Solo providers can be considered as safe, so we do not have to check if the user needs
+  // to provide an additional provider to login. An auth adapter with "solo" (like webauthn) means
+  // no "additional" auth needs to be provided to login (like OTP, MFA)
+  if (hasProvidedASoloProvider) return;
```
Single line return

dblythy commented on src/Auth.js:
```diff
@@ -298,11 +309,154 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+      }
+    }
+  });
+  if (hasProvidedAtLeastOneAdditionalProvider || !additionProvidersNotFound.length) return;
```
Single line return

dblythy commented on src/Auth.js:
```diff
@@ -298,11 +309,154 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+    Object.keys(authData).sort(),
+    async (acc, provider) => {
+      if (authData[provider] === null) {
+        acc.authData[provider] = null;
```
```
 if (!authData[provider]) {
        delete acc.authData[provider]
```

mtrezza replied:
Actually `authData[provider] === null` looks ood; this will equate to `true` for `undefined`, do we really want that?

Moumouls replied:
we need to use this to keep the client-provided authData structure. Setting authData to null allows removing authData from the database.

dblythy commented on src/RestWrite.js:
```diff
@@ -478,106 +448,157 @@ RestWrite.prototype.filteredObjectsByACL = function (objects) {
+// is not currently duplicating auth data ID
+RestWrite.prototype.ensureUniqueAuthDataId = async function () {
+  if (this.className !== '_User') return;
+  if (!this.data.authData) return;
```
Single line return

mtrezza replied:
Good catch and good candidate for lint rule!

dblythy replied:
Agree!

Moumouls replied:
@dblythy  not sur to understand the "single line" comment. You want a single "if" or i need to add brackets ?

dblythy replied:
I think the preferred styling should be to always have curly brackets whenever there’s an if statement

dblythy commented on src/RestWrite.js:
```diff
@@ -478,106 +448,157 @@ RestWrite.prototype.filteredObjectsByACL = function (objects) {
+      throw new Parse.Error(Parse.Error.ACCOUNT_ALREADY_LINKED, 'this auth is already used');
+    }
+
+    this.storage['authProvider'] = Object.keys(authData).join(',');
```
prefer dot notation

dblythy commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +478,94 @@ export class UsersRouter extends ClassesRouter {
+      if (typeof authData !== 'object')
+        throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'authData should be an object.');
+      if (user)
+        throw new Parse.Error(
```
single line throw

dblythy commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +478,94 @@ export class UsersRouter extends ClassesRouter {
+        } = authAdapter;
+        if (typeof challenge === 'function') {
+          acc[provider] =
+            (await challenge(
```
can we make this more readable

dblythy reviewed (commented):
Looks good! Just some minor nits, feel free to ignore

dblythy commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,47 @@ describe('Parse.User testing', () => {
+      },
+      getAuthType: function () {
+        return 'shortLivedAuth';
+      },
```
Prefer `object-shorthand`

dblythy commented:
On the topic of upgrading auth adapters, should we move auth adapters to their own package? Otherwise we could have the issue of new adapters requiring more and more dependancies.

For example, this adapter adds `@simplewebauthn/server`, and the OTP adapter will most likely add a TOTP library. It would be good to only have those dependancies installed if absolutely necessary.

Moumouls commented:
thanks @dblythy  for your review, i'll take a look to your feedbacks :) 
And yes, we need to move auth adapters out of the parse-server code base in a separated issue 👍

Moumouls commented:
@mtrezza  @dblythy, requested changes done. We just have a failing test about the Game Center, but otherwise everything is okay.

I added a test and slightly updated RestWrite.js in https://github.com/parse-community/parse-server/pull/7079/commits/20a32d5034918abb0479506612a9ef5d5ffd849b

I found an issue with my production apps. It's now covered.

Moumouls commented:
@dblythy  @mtrezza @davimacedo, the branch is now fully ready, squashed, up to date, tested.

Note: @mtrezza  i saw that you merged alpha but something went wrong with the commit history. This why i squashed the whole branch.

I hope we can merge this one asap (it has been waiting for a long time ahhaa)

mtrezza commented:
Could you please make sure all open comments are resolved and then request a review from the server-review team.

Moumouls commented:
@mtrezza  all comments resolved; branch up to date. review. asked the review of server-review team

mtrezza commented:
Alright, since @dblythy had a lot of feedback, maybe @dblythy has the time to review again?

dblythy commented on DEPRECATIONS.md:
```diff
@@ -9,6 +9,7 @@ The following is a list of deprecations, according to the [Deprecation Policy](h
 | DEPPS3 | Config option `enforcePrivateUsers` defaults to `true` | [#7319](https://github.com/parse-community/parse-server/pull/7319)   | 5.0.0 (2022)                    | 6.0.0 (2023)                    | deprecated            | -     |
 | DEPPS4 | Remove convenience method for http request `Parse.Cloud.httpRequest` | [#7589](https://github.com/parse-community/parse-server/pull/7589)   | 5.0.0 (2022)                    | 6.0.0 (2023)                    | deprecated            | -     |
 | DEPPS5 | Config option `allowClientClassCreation` defaults to `false` | [#7925](https://github.com/parse-community/parse-server/pull/7925)   | 5.3.0 (2023)                    | 7.0.0 (2024)                    | deprecated            | -     |
```
I just realized these dates are wrong, I think it should be (5.3.0) 2022 and 7.0.0 (2023), as version 6 is 2022, and 7 2023 cc @mtrezza

mtrezza replied:
Yes, 5.3.0 is 2022, 6.0.0 is 2023, 7.0.0 is 2024. So only the 5.3.0 year should be changed. If unrelated to this PR, could you open a new PR (probably without a new issue and reference the original issue that added this line as reference)?

mtrezza replied:
Solved in #7992

dblythy commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+  /**
+   * Legacy usage, if provided it will be triggered when authData related to this provider is touched (signup/update/login)
+   * otherwise you should implement validateSetup, validateLogin and validateUpdate
+  @param authData: the client provided authData
```
Is this valid JSDOc syntax? Shouldn't it be:

`* @param {Object} authData the client provided authData`

mtrezza replied:
JSDoc syntax addressed separately

dblythy commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+   * otherwise you should implement validateSetup, validateLogin and validateUpdate
+  @param authData: the client provided authData
+  @param options: additional options
+  @param req: RestWrite instance with config/auth/data
```
Instead of using `RestWrite` instance, can we perhaps use `{@link Parse.Cloud.FunctionRequest}` or something that makes sense to newer developers

Moumouls replied:
Thanks @dblythy  you are completely right! I'm refactoring the code to replace `req: RestWrite` with the already known `request` from `beforeSave/afterSave`.

mtrezza replied:
@Moumouls fixed?

dblythy commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+  @param user: Parse.User instance if Parse.User found
+  @returns a promise that resolves, the resolved value will be handled by the server like:
+  - resolve undefined|void|{} parse server will save authData
+  - resolve { doNotSave: boolean, response: Object} parse server will do not save provided authData and send response to the client under authDataResponse
```
Prefer

`- resolve { doNotSave: boolean, response: Object} Provided authData will not be saved, and response will be sent to the client under authDataResponse`

mtrezza replied:
Is that proper JSDoc or should it be:

> @param {Whatever type} user The description.


Also param and return should begin capitalized:
> @returns A promise that resolves...

I'm not sure but maybe we should always write code in single backticks (\``code`\`), because that may also format properly as code in IDEs and doc compilers.

And the return type is free text, it should either be a proper type or a new @typedef to describe the return so the IDE can pick it up. Incorrect docs are often a painful developer experience.

Haven't looked at the other code docs in this PR, but they should probably be reviewed as well.

mtrezza replied:
JSDoc issues addressed in separate review comments

dblythy commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+  @returns a promise that resolves, the resolved value will be handled by the server like:
+  - resolve undefined|void|{} parse server will save authData
+  - resolve { doNotSave: boolean, response: Object} parse server will do not save provided authData and send response to the client under authDataResponse
+  - resolve { response: Object } parse server will save authData and send response to the client under authDataResponse
```
Prefer `Parse Server` instead of `parse server`

mtrezza replied:
That's actually not just preferred but the correct spelling.

mtrezza replied:
fixed

dblythy commented on src/Adapters/Auth/index.js:
```diff
@@ -59,21 +61,60 @@ const providers = {
+    }
+    throw new Parse.Error(
+      Parse.Error.OTHER_CAUSE,
+      'Adapter not ready, need to implement validateAuthData or (validateSetUp, validateLogin, validateUpdate)'
```
`Adapter is not configured. Implement validateAuthData, or validateSetUp, validateLogin and validateUpdate`

mtrezza replied:
> Implement validateAuthData, or validateSetUp, validateLogin and validateUpdate

Sounds confusing, and / or what?

Maybe (if that's technically correct):

> Implement either validateAuthData or all of the following: validateSetUp, validateLogin and validateUpdate.

Or you have a better wording idea

dblythy replied:
I like `Adapter is not configured. Implement either validateAuthData or all of the following: validateSetUp, validateLogin and validateUpdate`

mtrezza replied:
fixed

dblythy commented on src/Config.js:
```diff
@@ -117,6 +118,7 @@ export class Config {
     this.validateSecurityOptions(security);
     this.validateSchemaOptions(schema);
     this.validateEnforcePrivateUsers(enforcePrivateUsers);
+    this.validateallowExpiredAuthDataToken(allowExpiredAuthDataToken);
```
`validateAllowExpiredAuthDataToken`

mtrezza replied:
has been fixed

dblythy commented on src/Config.js:
```diff
@@ -134,6 +136,12 @@ export class Config {
     }
   }
 
+  static validateallowExpiredAuthDataToken(allowExpiredAuthDataToken) {
```
`validateAllowExpiredAuthDataToken`

mtrezza replied:
fixed

dblythy reviewed (changes requested):
Some minor nits (sorry for being picky). Left some formatting changes out because I guess it's not required until we have stricter `ESLint` rules. 

Nice job on this PR, it's going to be a very welcome improvement.

mtrezza reviewed (changes requested):
JSDoc needs to be fully reviewed before merging.

- I'm not sure but maybe we should always write code in single backticks (`code`), because that may also format properly as code in IDEs and doc compilers.

- Return type should not be free text, it should either be a proper type {} or a new @typedef to describe the return so the IDE can pick it up. Incorrect docs are often a painful developer experience.

Moumouls commented:
@mtrezza after some thinking here. I would like to suggest removing the "webauthn" adapter and its dependency from this PR.

The webauthn adapter helped a lot to design new interfaces. But webauthn is going to evolve a lot in the next months https://www.apple.com/newsroom/2022/05/apple-google-and-microsoft-commit-to-expanded-support-for-fido-standard/

I also think that until we have a solution to properly split the code base and move out Adapters like Auth, GraphQL, and Database, we should avoid merging new adapters with new dependencies.

I'll be happy to get your feedback on this ! 🙂 

Last note: Currently webauthn on client-side, is hard to use and needs a lot of work to make it easy to use for end users. I'm not sure if it will be really used by the community. Instead, an MFA (suggested by @dblythy  ) will be a lot easier to implement/user for developers in their apps.

mtrezza commented:
These are 2 topics:

- a) How valuable is the webauthn feature?
- b) Move auth adapters out to slim down the core code base

I have no data to answer (a), but if you feel that its implementation is so complex that it won't provide much value, then feel free to remove it.

Generally we try to avoid adding heavy dependencies to the core codebase. The same arguments apply as in https://github.com/parse-community/parse-server/issues/7979. We will do (b) at some point; until then it shouldn't be an obstacle for developing the auth interface and adding new auth adapters. For example, it's probably better to have the GraphQL dependency in Parse Server now - despite its weight - than not having it at all.

rhuanbarreto commented:
For me the biggest win for this PR is the improved security once you can choose if sensitive data will saved at the database. Maybe if we split the new interface for auth and the webauthn in 2 PRs it would be released fast.

mtrezza commented:
We always prefer multiple smaller PRs over one large PR for many reasons, it's a continuous feedback that we give. It makes review easier, reduces risk and shortens submission-to-merge time.

I would also be in favor of removing webauthn from the general improvements and merge only that PR, and if necessary submit a separate PR for webauthn, unless the feature doesn't provide much value as has been suggested.

Moumouls commented:
Got it @mtrezza, i'll create new branch and remove the webauthn adapter from this one

Moumouls commented:
Okay thanks, @dblythy for your suggestion about switching to the Parse Trigger Request interface, really good catch!
I also tried to do my best to improve JSDoc on auth adapter.
I removed webauthn.

I wait now your final feedbacks.
@mtrezza  @dblythy

mtrezza commented on DEPRECATIONS.md:
```diff
@@ -11,7 +11,7 @@ The following is a list of deprecations, according to the [Deprecation Policy](h
 | DEPPS6 | Auth providers disabled by default | [#7953](https://github.com/parse-community/parse-server/pull/7953)   | 5.3.0 (2022)                    | 7.0.0 (2024)                    | deprecated            | -     |
 | DEPPS7 | Remove file trigger syntax `Parse.Cloud.beforeSaveFile((request) => {})` | [#7966](https://github.com/parse-community/parse-server/pull/7966)   | 5.3.0 (2022)                    | 7.0.0 (2024)                    | deprecated            | -     |
-
+| DEPPS8 | Allow login with expired authData token | [#7079](https://github.com/parse-community/parse-server/pull/7079)   | 5.3.0 (2022)                    | 7.0.0 (2024)                    | deprecated            | -     |
```
Add empty line below

Moumouls replied:
For wording/typos/comments feedback please use the github commit suggestion feature.
Then I'll be able to easily approve your requested changes.

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -18,6 +20,16 @@ const responses = {
   microsoft: { id: 'userId', mail: 'userMail' },
 };
 
+// A simple wrapper to allow usage of
```
Remove comment

Moumouls replied:
please use github commit suggestion

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2170,3 +2234,1055 @@ describe('facebook limited auth adapter', () => {
+        return originalFn(...params);
+      }
+      // Second call is triggered after beforeSave. A developer can modify authData during beforeSave.
+      // To perform a determinist login, parse need to ensure uniqueness of the authData.id into the database.
```
What's a "determinist login"?

Rephrase without referring to "parse":

```
... uniqueness of `auth.id` needs to be ensured.
```

Moumouls replied:
Determinist login means that the login will not fail randomly. For example, if 2 users exist with the same authData.id, depending on the insertion order into the database, the login result could change.

for rephrase please use github commit suggestion

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2170,3 +2234,1055 @@ describe('facebook limited auth adapter', () => {
+      // Second call is triggered after beforeSave. A developer can modify authData during beforeSave.
+      // To perform a determinist login, parse need to ensure uniqueness of the authData.id into the database.
+      // A developer with a direct access to the database could break something and duplicate authData.id.
+      // In this case, if parse detect 2 matching users for a singe authData.id; login/register will be canceled.
```
Don't refer to "parse":

```
... if 2 matching users ... are detected
```

Moumouls replied:
please use github commit suggestion

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2170,3 +2234,1055 @@ describe('facebook limited auth adapter', () => {
+      // To perform a determinist login, parse need to ensure uniqueness of the authData.id into the database.
+      // A developer with a direct access to the database could break something and duplicate authData.id.
+      // In this case, if parse detect 2 matching users for a singe authData.id; login/register will be canceled.
+      // Promise.resolve([true, true]) simulate this case with 2 matching users.
```
> simulates this case

Moumouls replied:
please use github commit suggestion

mtrezza commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2170,3 +2234,1055 @@ describe('facebook limited auth adapter', () => {
+    await reconfigureServer({ auth: { modernAdapter } });
+    const user = new Parse.User();
+
+    // Signup
```
Remove this and the following single word comments

Moumouls replied:
please use github commit suggestion

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,42 @@ describe('Parse.User testing', () => {
-        }
-      );
+    await Parse.User._logInWith('shortLivedAuth', {});
+    // Simulate a remotely expired token (like a short lived one)
```
What's a "remotely expired token"?

Moumouls replied:
i do not introduced this comment but i believe that a "remotely expired token" is an expired token in a 3rd party database. So the 3rd party will return that the provided client token is expired and will not login the user.

mtrezza replied:
Why has this been resolved? We are both not sure what it means apparently? Needs rewording.

rhuanbarreto replied:
This discussion is irrelevant for merging the PR. This comment was not created or changed by this PR. @Moumouls just moved from "then" to "await". Please look at the whole code change and you will see the comment was already there.

mtrezza replied:
Didn't notice this was just moved, yes, irrelevant for this PR

Moumouls replied:
+1 for @rhuanbarreto comment. This comment is OOS (out of scope). I think it's now time to merge this one.

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,42 @@ describe('Parse.User testing', () => {
+    await Parse.User._logInWith('shortLivedAuth', {});
+    // Simulate a remotely expired token (like a short lived one)
+    // In this case, we want success as it was valid once.
+    // If the client needs an updated one, do lock the user out
```
Replace "one" with what you are referring to

Moumouls replied:
please use commit suggestion

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,42 @@ describe('Parse.User testing', () => {
+    defaultConfiguration.auth.shortLivedAuth.setValidAccessToken('token');
+    Parse.User._registerAuthenticationProvider(provider);
+    await Parse.User._logInWith('shortLivedAuth', {});
+    // Simulate a remotely expired token (like a short lived one)
```
What's "remotely expired"?

Moumouls replied:
https://github.com/parse-community/parse-server/pull/7079#discussion_r891450043

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,42 @@ describe('Parse.User testing', () => {
+    await Parse.User._logInWith('shortLivedAuth', {});
+    // Simulate a remotely expired token (like a short lived one)
+    // In this case, we want success as it was valid once.
+    // If the client needs an updated one, do lock the user out
```
Replace "one"

Moumouls replied:
please use commit suggestion

mtrezza commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+
+/**
+ * @interface ParseAuthResponse
+ * @property {Boolean} [doNotSave] If true, parse server will do not save provided authData.
```
Typo, the product name is "Parse Server".

Moumouls replied:
please use commit suggestion

mtrezza commented:
Note that my review did not include JSDoc syntax. That still needs to be done.

Moumouls commented:
thanks @mtrezza i answered your comments, during your reviews about comments/wording/typos please use the Github commit suggestion, it'll not take additional time and will save me precious time 🙂

mtrezza commented:
If you need any specific info to make the corrections, please let me now in the comments, or maybe others have suggestions for how to rephrase some of the comments - in principle they should just be easily understandable.

Some tips to save time in the future:
- copy/paste in-code comments in a word processor, it will auto-detect typos; maybe it can also be switched on in your IDE
- in-code comments should refer to technical components rather than brand names; e.g. a reference to "parse" is too vague which indicates it is likely an unnecessary or confusing reference; for external code docs (visible to developer) brand names should be spelled correctly, e.g. "Parse Server" instead of "parse-server" or similar
- use terms that are widely understood; "determinist login" (I assume you meant "deterministic login"?) is something not every developer may interpret the same way

Maybe we should add this to our contribution guide.

Moumouls commented:
Hi @mtrezza 

> deterministic login

Yes

> Maybe we should add this to our contribution guide.

We can add your suggestion to the contribution guide. I think it will not really fix the problem for new contributors, but we will be able to use it as a documentation reference in review comments.


Btw, could you please edit you comments using github commit suggestion ?

![suggestion-block](https://user-images.githubusercontent.com/27959372/172564993-5ca9dc80-affc-41ae-96da-4c282c12480d.png)

rhuanbarreto commented on DEPRECATIONS.md:
```diff
@@ -11,7 +11,7 @@ The following is a list of deprecations, according to the [Deprecation Policy](h
 | DEPPS6 | Auth providers disabled by default | [#7953](https://github.com/parse-community/parse-server/pull/7953)   | 5.3.0 (2022)                    | 7.0.0 (2024)                    | deprecated            | -     |
 | DEPPS7 | Remove file trigger syntax `Parse.Cloud.beforeSaveFile((request) => {})` | [#7966](https://github.com/parse-community/parse-server/pull/7966)   | 5.3.0 (2022)                    | 7.0.0 (2024)                    | deprecated            | -     |
-
+| DEPPS8 | Allow login with expired authData token | [#7079](https://github.com/parse-community/parse-server/pull/7079)   | 5.3.0 (2022)                    | 7.0.0 (2024)                    | deprecated            | -     |
```
```suggestion
| DEPPS8 | Allow login with expired authData token | [#7079](https://github.com/parse-community/parse-server/pull/7079)   | 5.3.0 (2022)                    | 7.0.0 (2024)                    | deprecated            | -     |

```

rhuanbarreto commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,42 @@ describe('Parse.User testing', () => {
+    await Parse.User._logInWith('shortLivedAuth', {});
+    // Simulate a remotely expired token (like a short lived one)
+    // In this case, we want success as it was valid once.
+    // If the client needs an updated one, do lock the user out
```
```suggestion
    // If the client needs an updated token, do lock the user out
```

rhuanbarreto commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,42 @@ describe('Parse.User testing', () => {
+    await Parse.User._logInWith('shortLivedAuth', {});
+    // Simulate a remotely expired token (like a short lived one)
+    // In this case, we want success as it was valid once.
+    // If the client needs an updated one, do lock the user out
```
```suggestion
    // If the client needs an updated token, do lock the user out
```

rhuanbarreto commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+
+/**
+ * @interface ParseAuthResponse
+ * @property {Boolean} [doNotSave] If true, parse server will do not save provided authData.
```
```suggestion
 * @property {Boolean} [doNotSave] If true, Parse Server will do not save provided authData.
```

rhuanbarreto reviewed (approved):
I added the changes as comments. Please LGTM!

rhuanbarreto commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -18,6 +20,16 @@ const responses = {
   microsoft: { id: 'userId', mail: 'userMail' },
 };
 
+// A simple wrapper to allow usage of
+// expectAsync().toBeRejectedWithError(errorMessage)
```
```suggestion
```

rhuanbarreto commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2170,3 +2234,1055 @@ describe('facebook limited auth adapter', () => {
+        return originalFn(...params);
+      }
+      // Second call is triggered after beforeSave. A developer can modify authData during beforeSave.
+      // To perform a determinist login, parse need to ensure uniqueness of the authData.id into the database.
```
```suggestion
      // To perform a determinist login, the uniqueness of `auth.id` needs to be ensured.
```

rhuanbarreto commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2170,3 +2234,1055 @@ describe('facebook limited auth adapter', () => {
+      // Second call is triggered after beforeSave. A developer can modify authData during beforeSave.
+      // To perform a determinist login, parse need to ensure uniqueness of the authData.id into the database.
+      // A developer with a direct access to the database could break something and duplicate authData.id.
+      // In this case, if parse detect 2 matching users for a singe authData.id; login/register will be canceled.
```
```suggestion
      // In this case, if 2 matching users are detected for a single authData.id, then the login/register will be canceled.
```

rhuanbarreto commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2170,3 +2234,1055 @@ describe('facebook limited auth adapter', () => {
+      // To perform a determinist login, parse need to ensure uniqueness of the authData.id into the database.
+      // A developer with a direct access to the database could break something and duplicate authData.id.
+      // In this case, if parse detect 2 matching users for a singe authData.id; login/register will be canceled.
+      // Promise.resolve([true, true]) simulate this case with 2 matching users.
```
```suggestion
      // Promise.resolve([true, true]) simulates this case with 2 matching users.
```

rhuanbarreto commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2170,3 +2234,1055 @@ describe('facebook limited auth adapter', () => {
+    await reconfigureServer({ auth: { modernAdapter } });
+    const user = new Parse.User();
+
+    // Signup
```
```suggestion
```

rhuanbarreto commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2170,3 +2234,1055 @@ describe('facebook limited auth adapter', () => {
+    await reconfigureServer({ auth: { modernAdapter }, allowExpiredAuthDataToken: false });
+    const user = new Parse.User();
+
+    // Signup
```
```suggestion
```

rhuanbarreto commented on spec/AuthenticationAdapters.spec.js:
```diff
@@ -2170,3 +2234,1055 @@ describe('facebook limited auth adapter', () => {
+    await user.save({ authData: { modernAdapter: { id: 'modernAdapter' } } });
+
+    expect(modernAdapter.validateSetUp).toHaveBeenCalledTimes(1);
+    // Login
```
```suggestion
```

rhuanbarreto reviewed (approved):
Another batch of github suggestions. Please let's help the developer on the review!

Moumouls commented:
Huge thanks @rhuanbarreto , it took me less that a minute to apply changes 🚀  Thanks for your contribution 👍

mtrezza commented on src/Adapters/Auth/index.js:
```diff
@@ -61,19 +62,59 @@ const providers = {
+        if (hasAuthDataConfigured) {
+          return adapter.validateUpdate(authData, options, requestObject, req.config);
+        }
+        // Let's setup if the user does not have the provider configured
```
```suggestion
        // Set up if the user does not have the provider configured
```

mtrezza reviewed (changes requested):
Great work, we're getting closer to merge

mtrezza commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+ * @interface ParseAuthResponse
+ * @property {Boolean} [doNotSave] If true, Parse Server will do not save provided authData.
+ * @property {Object} [response] If set, parse server will send the provided response to the client under authDataResponse
+ * @property {Object} [save] If set, parse server will save the object provided into this key, instead of client provided authData
```
> Parse Server

check JSDoc syntax

dblythy replied:
```suggestion
 * @property {Object} [save] If set, Parse Server will save the object provided into this key, instead of client provided authData
```

mtrezza commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+    this.policy = 'default';
+  }
+  /**
+   * @param appIds The specified app ids in the configuration
```
> app IDs

check JSDoc syntax

Moumouls replied:
```suggestion
   * @param appIds The specified app IDs in the configuration
```

mtrezza commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+   * @param {Object} authData The client provided authData
+   * @param {Object} options Additional options
+   * @param {Parse.Cloud.TriggerRequest} request
+   * @param {Object} config
```
Missing object description for `config`

check JSDoc syntax

Moumouls replied:
```suggestion
   * @param {Object} config Parse Server config object
```

mtrezza commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+   * @param {Object} authData The client provided authData
+   * @param {Object} options Additional options
+   * @param {Parse.Cloud.TriggerRequest} request
+   * @param {Object} config
```
check JSDoc syntax

Moumouls replied:
```suggestion
   * @param {Object} config Parse Server config object
```

mtrezza commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+
+  /**
+   * Triggered when user provide authData related to this provider
+   * he is not logged in and has already set this provider before
```
Rephrase... "he"?

check JSDoc syntax

dblythy replied:
```suggestion
   * The user is not logged in and has already set this provider before
```

mtrezza commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+
+  /**
+   * Triggered when user provide authData related to this provider
+   * he is logged in and has already set this provider before
```
Rephrase... "he"?

check JSDoc syntax

Moumouls replied:
```suggestion
   * the user is logged in and has already set this provider before
```

mtrezza commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+   * @param {(Object|undefined)} authData Auth data provided by the client, can be used for validation
+   * @param {Object} options Additional options
+   * @param {Parse.Cloud.TriggerRequest} request
+   * @param {Object} config
```
> pre-authentication

check JSDoc syntax

Moumouls replied:
```suggestion
   * @param {Object} config Parse Server config object
```

mtrezza reviewed (changes requested):
Minor issues in the JSDoc

dblythy commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,42 @@ describe('Parse.User testing', () => {
+        id: '12345',
+        access_token: 'token',
+      },
+      restoreAuthentication: function () {
```
```suggestion
      restoreAuthentication() {
```

dblythy commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,42 @@ describe('Parse.User testing', () => {
+      restoreAuthentication: function () {
+        return true;
+      },
+      deauthenticate: function () {
```
```suggestion
      deauthenticate() {
```

dblythy commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,42 @@ describe('Parse.User testing', () => {
+      deauthenticate: function () {
+        provider.authData = {};
+      },
+      authenticate: function (options) {
```
```suggestion
      authenticate(options) {
```

dblythy commented on spec/ParseUser.spec.js:
```diff
@@ -1813,22 +1870,42 @@ describe('Parse.User testing', () => {
+      authenticate: function (options) {
+        options.success(this, provider.authData);
+      },
+      getAuthType: function () {
```
```suggestion
      getAuthType() {
```

dblythy commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+/**
+ * @interface ParseAuthResponse
+ * @property {Boolean} [doNotSave] If true, Parse Server will do not save provided authData.
+ * @property {Object} [response] If set, parse server will send the provided response to the client under authDataResponse
```
```suggestion
 * @property {Object} [response] If set, Parse Server will send the provided response to the client under authDataResponse
```

dblythy commented on src/Adapters/Storage/Postgres/PostgresStorageAdapter.js:
```diff
@@ -1308,15 +1307,21 @@ export class PostgresStorageAdapter implements StorageAdapter {
         return;
       }
       var authDataMatch = fieldName.match(/^_auth_data_([a-zA-Z0-9_]+)$/);
+      const authDataAlreadyExists = !!object['authData'];
```
```suggestion
      const authDataAlreadyExists = !!object.authData;
```

dblythy commented on src/Adapters/Storage/Postgres/PostgresStorageAdapter.js:
```diff
@@ -1308,15 +1307,21 @@ export class PostgresStorageAdapter implements StorageAdapter {
       }
 
       columnsArray.push(fieldName);
+
```
```suggestion
```

dblythy commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +491,106 @@ export class UsersRouter extends ClassesRouter {
+      const results = await Auth.findUsersWithAuthData(req.config, authData);
+
+      try {
+        if (!results[0] || results.length > 1)
```
```suggestion
        if (!results[0] || results.length > 1) {
```

dblythy commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +491,106 @@ export class UsersRouter extends ClassesRouter {
+      try {
+        if (!results[0] || results.length > 1)
+          throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'User not found.');
+
```
```suggestion
}
```

dblythy reviewed (changes requested):
Just some minor nits again

Moumouls commented:
Thanks, @dblythy  for your commits suggestions!

@mtrezza I commented some "commit suggestions" about your review. Also again, for this kind of feedback please use "Commit Suggestion" Github feature.

@dblythy and @rhuanbarreto helped a lot, and I also saw that @dblythy even suggested some commits on your review.
As personal feedback, I can say that it's quite frustrating since commit suggestions are easy to use or there are maybe some details that I don't know/understand...

btw thanks @rhuanbarreto @dblythy

mtrezza commented:
@Moumouls A reviewer may or may not contribute with code changes to a PR, *in addition to the review itself*.

Please refrain from pressuring others to contribute to your PR by repeatedly demanding to write the actual changes in addition to their review. As you well know, the review feedback is publicly visible and anyone can make commit suggestions based on other feedback.

Moumouls commented:
Any additional suggestions?

mtrezza commented:
See https://github.com/parse-community/parse-server/pull/7079#pullrequestreview-1000520125, the JSDoc issues are still open.

mtrezza commented on src/Adapters/Auth/AuthAdapter.js:
```diff
@@ -1,20 +1,102 @@
+ * default: can be combined with ONE additional auth provider if additional configured on user
+ * additional: could be only used with a default policy auth provider
+ * solo: Will ignore ALL additional providers if additional configured on user
+ * @typedef {"default" | "additional" | "solo"} AuthPolicy
```
Declare this new type so it can be referenced properly, error checked and provide auto-completion:

```js

/**
 * The auth policy.
 * 
 * @typedef {'default' | 'additional' | 'solo'} AuthPolicy
 */
const AuthPolicy = Object.freeze({
  /** Auth provider can be combined with one additional auth provider, if there is an additional auth provider configured for the user. */
  default: 'default',
  /** ... */
  additional: 'additional',
  /** ... */
  solo: 'solo',
});

class AuthAdapter {
  /**
   * @param {AuthPolicy} policy The auth policy to use for the adapter.
   */
  constructor(policy = AuthPolicy.default) {
    if (![...Object.values(AuthPolicy)].includes(policy)) {
      // error
    }
  }
}
```

Moumouls commented:
I detected a bug in my advanced usage of this PR. It's now fixed and covered.

mtrezza commented:
@parse-community/server could anyone commit the few remaining bits from the review comments? So we can get this merged.

dblythy commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +491,106 @@ export class UsersRouter extends ClassesRouter {
+      if (user) {
+        throw new Parse.Error(
+          Parse.Error.OTHER_CAUSE,
+          'You cant provide username/email and authData, only use one identification method.'
```
```suggestion
          'You cannot provide username/email and authData, only use one identification method.'
```

dblythy commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +491,106 @@ export class UsersRouter extends ClassesRouter {
+      if (Object.keys(authData).filter(key => authData[key].id).length > 1) {
+        throw new Parse.Error(
+          Parse.Error.OTHER_CAUSE,
+          'You cant provide more than one authData provider with an id.'
```
```suggestion
          'You cannot provide more than one authData provider with an id.'
```

dblythy commented:
Looks mostly good - will run over again today

mtrezza commented:
I guess https://github.com/parse-community/parse-server/pull/7079#discussion_r895222394 cannot be easily suggested via review, so I hope @Moumouls can add this to the PR.

Moumouls commented:
@mtrezza @dblythy , I'll not add https://github.com/parse-community/parse-server/pull/7079#discussion_r895222394

The AuthAdapter file is only used for documentation/example purposes. Also as you can see, all other adapters are in fact not classes, there is no notion of constructor and parameter validation. The `policy` key is just an optional key that a developer could add (or not add) to the adapter object to tell Parse Server how to handle the auth.

Anyway, I added validation on the policy key. it could help a developer to debug an auth adapter !

mtrezza commented:
> The policy key is just an optional key that a developer could add (or not add) to the adapter object to tell Parse Server how to handle the auth.

How would the developer add that key to the adapter?

dblythy commented on src/RestQuery.js:
```diff
@@ -659,7 +659,7 @@ RestQuery.prototype.runFind = function (options = {}) {
       if (this.className === '_User' && !findOptions.explain) {
         for (var result of results) {
-          cleanResultAuthData(result);
+          this.cleanResultAuthData(result, this.auth, this.config);
```
What's the purpose of these arguments?

dblythy commented on src/RestWrite.js:
```diff
@@ -487,106 +449,161 @@ RestWrite.prototype.filteredObjectsByACL = function (objects) {
+    key => this.data.authData[key] && this.data.authData[key].id
+  );
+
+  if (!hasAuthDataId) return;
```
```suggestion
  if (!hasAuthDataId) {
    return;
  }
```

dblythy commented on src/RestWrite.js:
```diff
@@ -487,106 +449,161 @@ RestWrite.prototype.filteredObjectsByACL = function (objects) {
+
+  if (!hasAuthDataId) return;
+
+  const r = await Auth.findUsersWithAuthData(this.config, this.data.authData);
```
Poorly named variable

dblythy commented on src/RestWrite.js:
```diff
@@ -487,106 +449,161 @@ RestWrite.prototype.filteredObjectsByACL = function (objects) {
+};
+
+RestWrite.prototype.handleAuthData = async function (authData) {
+  const r = await Auth.findUsersWithAuthData(this.config, authData);
```
Poorly named variable

dblythy commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +491,106 @@ export class UsersRouter extends ClassesRouter {
+      } catch (e) {
+        // Rewrite the error to avoid guess id attack
+        logger.error(e);
+        throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'User not found.');
```
Should the error here be ignored here? Could the error be thrown from `await validator`

dblythy commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +491,106 @@ export class UsersRouter extends ClassesRouter {
+      Object.keys(challengeData).sort(),
+      async (acc, provider) => {
+        const authAdapter = req.config.authDataManager.getValidatorForProvider(provider);
+        if (!authAdapter) return acc;
```
```suggestion
        if (!authAdapter) {
          return acc;
        }
```

dblythy commented on src/Adapters/Auth/index.js:
```diff
@@ -121,12 +176,12 @@ module.exports = function (authOptions = {}, enableAnonymousUsers = true) {
-
-    return authDataValidator(adapter, appIds, providerOptions);
+    const authAdapter = loadAuthAdapter(provider, authOptions);
+    if (!authAdapter) return;
```
```suggestion
    if (!authAdapter) {
      return;
    }
```

dblythy commented on src/Auth.js:
```diff
@@ -307,11 +320,166 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+};
+
+const hasMutatedAuthData = (authData, userAuthData) => {
+  if (!userAuthData) return { hasMutatedAuthData: true, mutatedAuthData: authData };
```
```suggestion
  if (!userAuthData) {
    return { hasMutatedAuthData: true, mutatedAuthData: authData };
  }
```

dblythy commented on src/Auth.js:
```diff
@@ -307,11 +320,166 @@ Auth.prototype._getAllRolesNamesForRoleIds = function (roleIDs, names = [], quer
+  const mutatedAuthData = {};
+  Object.keys(authData).forEach(provider => {
+    // Anonymous provider is not handled this way
+    if (provider === 'anonymous') return;
```
```suggestion
    if (provider === 'anonymous') {
      return;
    }
```

dblythy commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +491,106 @@ export class UsersRouter extends ClassesRouter {
+      user = await this._authenticateUserFromRequest(req);
+    }
+
+    if (!challengeData) throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'Nothing to challenge.');
```
```suggestion
    if (!challengeData) {
      throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'Nothing to challenge.');
    }
```

dblythy commented on src/Routers/UsersRouter.js:
```diff
@@ -453,6 +491,106 @@ export class UsersRouter extends ClassesRouter {
+
+    if (!challengeData) throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'Nothing to challenge.');
+
+    if (typeof challengeData !== 'object')
+      throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'challengeData should be an object.');
```
```suggestion
    if (typeof challengeData !== 'object') {
      throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'challengeData should be an object.');
    }
```

dblythy reviewed (changes requested):
- I haven't seen any tests for how a developer would throw a custom error from an auth adapter, such as `challenge`.

- I have also found that if an error is thrown from validateSetUp, validateUpdate, it is not returned to the client or caught by the server. Errors thrown from `challenge` are returned as `{"code":1,"message":"Internal server error."}`.

- I noticed the last argument in all the adapter functions is a `Config` option, which is inconsistent with the Parse Cloud protocol, and pending discussion in #7869

- The request param (`Parse.Cloud.TriggerRequest`) field `triggerName` is always empty, it would be good to pass `authAdapter.field` to there (e.g `challengeAdapter.challenge`).

- The request param (`Parse.Cloud.TriggerRequest`) field `user` is always empty, I think it's intuitive for the user calling the auth to be passed in here (can other users' with write access update authData on other users?)

- The third argument for `challenge` and the second argument for the other functions is set to JSDocs as "additional options", but in my usage it shows up as the Auth adapter itself (perhaps it should be "this", but then what's the point of the argument):

```js
{
      validateSetUp: [Function: validateSetUp],
      validateUpdate: [Function: validateUpdate],
      validateLogin: [Function: validateLogin],
      validateAppId: [Function: validateAppId],
      validateAuthData: [Function: validateAuthData],
      challenge: [Function: challenge],
      options: [Object]
},

```


- Throwing from `validateAuthData` returns `User not found` regardless of the error

- I think error code "OTHER CAUSE" is overused here, I think we should probably get some new error codes considering this is an Auth adapter upgrade. Other cause is code `-1` which I think should probably only be used sparingly as a last resort. Your thoughts @mtrezza?

- Perhaps the new Auth Adapter protocol should get its own spec - the spec file this in is quite bloated and more or less refers to custom auth _providers_.

- I also think the tests should start with a simple test which maps out usage (no spies, expects on validateAppId fields for example) so developers can get an easy view of how to implement. All the existing tests are quite complex to cover the code - a few simple ones at the start e.g how [Parse Cloud tests](https://github.com/parse-community/parse-server/blob/alpha/spec/CloudCode.spec.js#L42) are done could be nice).

- It would be good for an enum on `policy` so it's clear what "default", "solo" and "additional" actually mean

- I like how you don't have to return promises - nice improvement.

Thanks again for your hard work here @Moumouls!

mtrezza commented:
> I think error code "OTHER CAUSE" is overused here, I think we should probably get some new error codes considering this is an Auth adapter upgrade. Other cause is code -1 which I think should probably only be used sparingly as a last resort. Your thoughts @mtrezza?

Yes

mtrezza commented:
@Moumouls, @rhuanbarreto, any feedback to @dblythy's points?

rhuanbarreto commented:
My only feedback is that this PR will never be merged once many discussions are outside the scope of the PR, like correct usage of error messages. I think its functionality should be split in many different PRs with features.

In this PR, the only feature I'm interested in is the ability to not save sensitive data (authData) in the database, once this is a security issue. If the database leaks, not only the session token but also auth data gets leaked which can lead to a bigger liability due to GDPR.

mtrezza commented:
> discussions are outside the scope of the PR

Which of the points in https://github.com/parse-community/parse-server/pull/7079#pullrequestreview-1027396179 do you see outside of the scope of this PR? Let's see what we can split off into separate PRs.

dblythy commented:
> like correct usage of error messages

Our community has been very vocal on breaking changes, and if we introduce new features only to potentially break them (such as the expected error codes - e.g, a developer builds an auth system around the error `OTHER_CAUSE` for it to only change to a different error one version later) that’s only going to cause more frustration and delays in line with our policies. 

This PR is _almost_ there in my view though and most of my feedback is around future proofing / making tests easier to interpret.

Moumouls commented:
Hi @mtrezza, @rhuanbarreto @dblythy 

Yes, I tried to do my best in this PR, but then I start to "waste" time trying to make these new features join the official Parse Server repo (like some other features).

This PR helped a lot, my team, to better control the auth, implement reusable 2FA, MFA, and improve the security of data saved in the database.

Parse Server auth system needed a major refactor. Maybe this PR could have been split into many tiny PR (the future will tell us maybe if a contributor wants to try), but a global overview was also needed to perform this refactor.

I tried to find the best time/effort/value to produce these changes. But for now, I'll not provide updates to this PR. And I'll not probably sync new changes that we do internally with my team to continue to improve Parse Auth.

I always try to push new features to Parse Server, like Defined Schema, Parse Server config in the request object, Mongo Serverless support, and a new auth system, but even for a "frequent" committer, it feels harder and harder to contribute to Parse Server.

I've many plans to improve Parse Server, but based on my current "feeling" and how things going on here I'll prefer to preserve my effort for my team and maybe start a complete fork (Yarn Monorepo, Typescript, Serverless goal, Terraform templates to run Parse Server in 100% serverless on GCP, focus on GraphQL, Test helper package for Jest, improve GraphQL API performance, Improve Parse Server query performance, A Job task adapter to send Jobs to services like Google Cloud Tasks).

I really want to send all of this to the community repo but it feels nearly impossible when I see how this PR is going.

Have a nice day all !

mtrezza commented:
Complex changes require more scrutiny.

Everybody is thankful that you share your code and you surely invested a lot of time writing this change. Let's be mindful that others invested a lot of time as well to review your change. Review feedback is an important part of the open-source process to uphold code quality and we want to be thankful to everyone involved as their time is as valuable as yours.

Not every feedback needs to be incorporated. Some feedback may be ideas to inspire follow-up contributions while other feedback may be a requirement to merge. Then there is feedback in between where the change works as is but merging it likely causes complications later on. Evaluating these in a mindful discussion is what everyone hopefully aims for.

We have seen a similar pattern in https://github.com/parse-community/parse-server/pull/7091 where you submitted a change but didn't have the capacity to incorporate the review feedback. Someone else then continued your PR in https://github.com/parse-community/parse-server/pull/7418. That's legitimate in a collaborative environment.

So let's better invest our time to look at the [detailed feedback](https://github.com/parse-community/parse-server/pull/7079#pullrequestreview-1027396179) point by point and determine:
- what is an idea for a future improvment
- what is essential to fix before merge
- what is optional but would be highly beneficial before merge

mtrezza commented:
Continued in https://github.com/parse-community/parse-server/pull/8156.
